src/realm/object-store/sync/sync_session.cpp

```diff
@@ -326,24 +332,23 @@ SyncSession::handle_refresh(const std::shared_ptr<SyncSession>& session, bool re
 
 SyncSession::SyncSession(SyncClient& client, std::shared_ptr<DB> db, const RealmConfig& config,
                          SyncManager* sync_manager)
-    : m_config(config)
-    , m_db(std::move(db))
-    , m_flx_subscription_store([this](bool use_flx_sync) -> std::shared_ptr<sync::SubscriptionStore> {
-        if (!use_flx_sync) {
-            return nullptr;
-        }
-
-        return sync::SubscriptionStore::create(m_db, [this](int64_t new_version) {
-            util::CheckedLockGuard lk(m_state_mutex);
-            if (m_state != State::Active && m_state != State::WaitingForAccessToken) {
-                return;
-            }
-            // There may be no session yet (i.e., waiting to refresh the access token).
-            if (m_session) {
-                m_session->on_new_flx_sync_subscription(new_version);
-            }
+    : m_config{config}
+    , m_db{std::move(db)}
+    , m_flx_subscription_store{}
+    , m_original_sync_config{m_config.sync_config}
+    , m_migration_store{[&]() {
+        using MigrationState = sync::MigrationStore::MigrationState;
+        return sync::MigrationStore::create(m_db, [this](MigrationState) {
+            // Migration state changed - Update the configuration to
+            // match the new sync mode.
+            util::CheckedLockGuard cfg_lock(m_config_mutex);
+            m_config.sync_config = m_migration_store->convert_sync_config(m_original_sync_config);
+
+            // If using FLX, set up m_flx_subscription_store and the history_write_validator
+            util::CheckedLockGuard state_lock(m_state_mutex);
```

Comment:
Why do we need to hold the state mutex here? Isn't the subscription store guarded by the config mutex now?